File: Common_Concepts/binary_search.rb
class Array
  def my_bsearch(num)
    return nil if self.length == 0

    mid = self.length / 2
    if self[mid] == num
      return mid
    elsif self[mid] > num
      return self[0...mid].my_bsearch(num)
    else
      new_id = self[mid + 1..-1].my_bsearch(num)
      return new_id == nil ? nil : mid + new_id + 1
    end
  end
end

p [1, 3, 4, 5, 10, 11].my_bsearch(1)
p [1, 3, 4, 5, 10, 11].my_bsearch(3)
p [1, 3, 4, 5, 10, 11].my_bsearch(4)
p [1, 3, 4, 5, 10, 11].my_bsearch(5)
p [1, 3, 4, 5, 10, 11].my_bsearch(10)
p [1, 3, 4, 5, 10, 11].my_bsearch(11)
p [1, 3, 4, 5, 10, 11].my_bsearch(6)
p [1, 3, 4, 5, 10, 11].my_bsearch(0)

File: Common_Concepts/binarySearch.js
const bSearch = (arr, num) => {
  if (arr.length == 0) return null;

  let mid = Math.floor(arr.length / 2);
  if (arr[mid] == num) {
    return mid;
  } else if (arr[mid] > num) {
    return bSearch(arr.slice(0, mid), num);
  } else {
    let newId = bSearch(arr.slice(mid + 1), num);
    return (newId === null ? null : mid + 1 + newId);
  }
};

let arr1 = [1, 4, 9, 10, 19, 21];
console.log(bSearch(arr1, 1));
console.log(bSearch(arr1, 4));
console.log(bSearch(arr1, 9));
console.log(bSearch(arr1, 10));
console.log(bSearch(arr1, 19));
console.log(bSearch(arr1, 21));
console.log(bSearch(arr1, 20));
console.log(bSearch(arr1, 2));
